corehq/motech/fhir/tests/test_api_endpoints.py

```diff
@@ -0,0 +1,290 @@
+        clear_plan_version_cache()
+        super().tearDownClass()
+
+    def test_search(self):
+        # Tests filtering by a non-indexed case property
+        url = get_endpoint_url(BASE_URL, '/Patient?birthdate=1990-01-01')
+        response = requests.get(url, auth=(API_USERNAME, API_PASSWORD))
+
+        # Some things will be different between the expected search
+        # result and the actual response. Use `json_diff()` to confirm
+        # expected differences, and that everything else is the same.
+        diffs = sorted([diff.path for diff in json_diff(response.json(),
+                                                        get_search_result())])
+        self.assertEqual(diffs, ['id', 'meta.lastUpdated'])
+
+    def test_get(self):
+        url = get_endpoint_url(BASE_URL, f'/Patient/{FOO_CASE_ID}')
+        response = requests.get(url, auth=(API_USERNAME, API_PASSWORD))
+        self.assertEqual(response.json(), FOO_PATIENT)
+
+    def test_auth_bad_password(self):
+        # Authentication should use the same code as existing API
+        url = get_endpoint_url(BASE_URL, f'/Patient/{FOO_CASE_ID}')
+        with self.assertRaisesRegex(requests.HTTPError, '[Ff]orbidden'):
+            requests.get(url, auth=(API_USERNAME, 'bad_password'))
+
+    def test_auth_bad_username(self):
+        url = get_endpoint_url(BASE_URL, f'/Patient/{FOO_CASE_ID}')
+        with self.assertRaisesRegex(requests.HTTPError, '[Ff]orbidden'):
+            # Error should be the same as for bad password: It should
+            # not reveal that the user does not exist.
+            requests.get(url, auth=('admin@example.com', API_PASSWORD))
+
+    def test_auth_bad_resource(self):
+        url = get_endpoint_url(BASE_URL, f'/Patient/{BAZ_CASE_ID}')
+        with self.assertRaisesRegex(requests.HTTPError, '[Nn]ot found'):
+            requests.get(url, auth=(API_USERNAME, API_PASSWORD))
+
+
+def create_person_cases(owner_id):
+    """
+    Creates cases for Fred Foo and Barbara Bar.
+    """
+    different_owner_id = uuid4().hex
+    submit_case_blocks([
+        get_foo_caseblock(owner_id).as_text(),
+        get_bar_caseblock(owner_id).as_text(),
+        get_baz_caseblock(different_owner_id).as_text(),
+    ], DOMAIN)
+
+
+def get_search_result():
+    bundle_id = str(uuid4())
+    return {
+        "resourceType": "Bundle",
+        "id": bundle_id,
+        "meta": {
+            "lastUpdated": "2021-01-14T07:54:36.100+00:00"
+        },
+        "type": "searchset",
+        "total": 2,
+        "link": [
+            {
+                "relation": "self",
+                "url": get_endpoint_url(BASE_URL, '/Patient?birthdate=1990-01-01'),
+            },
+            # Keeping this here for reference. This is how pagination
+            # works by default in HAPI FHIR. We can use any GET params
+            # way we want, as long as the URL we provide here will
+            # return a Bundle with the next page of search results.
```

Comment:
HQ needs to keep a track of the bundle ids? Confused about the bundle ID being present in the "next" request.

Reply:
Good question!

We don't need to do pagination the same way as HAPI FHIR.

But, if we give bundles an ID, then page 2 of the same bundle should have the same ID.

Luckily, is seems from [the spec](https://www.hl7.org/fhir/bundle.html) that `id` is an optional attribute! It might be simplest just to leave it out.

Reply:
Cool. I don't see the purpose of it in the request as of now. At least not from HQ perspective, and for the side rendering the search result, they might use it to keep track of multiple requests instead of using the params I guess. If the logic is to
1. return the same bundle id if present in the request params
2. else generate a new one
that seems totally okay.

Reply:
I see, so then we should return if present, else return nothing for bundle ID.

> Is that OK? How do we deal with this in HQ in other places?
Isn't that a risk with any pagination?

Reply:
For pagination where we are doing a search for each page, yes. The alternative is to search, and just return case_type-case_id pairs. Then save or cache those. For each requested page, fetch the full cases where the IDs are in that page. In other words, only search once.

I think this is what HAPI FHIR is doing. It is saving the resultset of IDs, and that's what the bundle ID is. If we were to do it this way, I would drop all resultsets after, say, a day.

Reply:
`case_type-case_id pairs`
That is what we are doing in tests already right? Search returns IDs and URLs to fetch details from for each entry.

You mean cache by params?
I see, caching would still give back the same result but it would then be returning outdated records? Sorry its not clear how caching would solve the issue with records getting skipped.

Reply:
> You mean cache by params

Yes. I mean something like this:
```python
@quickcache(['domain', 'params'], timeout=1 * 60 * 60)
def search(domain, params):
    xpath = _build_xpath(params)
    search = CaseSearchES().domain(domain).xpath_query(domain, xpath)
    results = search.run()
    return [(r['_id'], r['case_type']) for r in results.raw_hits]
```

So for one hour we can get the same cases using the same search params without hitting the DB. For each page, we fetch the latest case data (even if it changes after the search)

Reply:
> For each page, we fetch the latest case data (even if it changes after the search

Confused by this. Won't it be stale data since we would returning the cached data, in case there has been any update or removal of a case within that data set returned for a page with specific params?

Reply:
As you know, [fetching a list of IDs](https://github.com/dimagi/commcare-hq/blob/08eec95e4246bf8eea78805a16a969b5a905f4d4/corehq/form_processor/interfaces/dbaccessors.py#L423) and then iterating them ([often in chunks](https://github.com/dimagi/commcare-hq/blob/85f8d7f62627fd2a39e87c8461c572986684f81f/corehq/apps/domain/deletion.py#L171)) to get instances is a common pattern in HQ. This is the same.

Returning the latest state of the case, even if its state has changed since its ID was fetched, is expected behaviour (at least among devs :upside_down_face: ) I think. That is what [Data Forwarding does](https://github.com/dimagi/commcare-hq/blob/f6a46d653ef7e866175218c3ef13a06ffbf6df78/corehq/motech/repeaters/models.py#L599), and what this would do too.

You make a good point that the case could have been deleted in the hour that the list of IDs is cached. We should be explicit about what we do with this. Just as deleted cases have a "deleted" property set to `True`, or their "doc_type" property is set to "CommCareCase-Deleted", maybe we can represent that state with the "[active](https://www.hl7.org/fhir/patient-definitions.html#Patient.active)" attribute set to `False`? Or we can return an empty JSON object?

Reply:
:thinking: An alternative might be to create a partitioned table for bundles, like [SyncLogSQL](https://github.com/dimagi/commcare-hq/blob/599b34bb79f9ebf9886072e39289894ef3cefbd2/corehq/ex-submodules/casexml/apps/phone/models.py#L431-L432), but partition by day instead of week, and then [drop bundles](https://github.com/dimagi/commcare-hq/blob/0008337e32ed08c11dc04f043bd742739f48dfc4/corehq/ex-submodules/casexml/apps/phone/tasks.py#L94) after a day or two.